### PR TITLE
Feature: salt-masterless formulas

### PR DESCRIPTION
**Pull request by acornies**
This PR is a new feature for the salt-masterless provisioner.

Use case: The salt-masterless provisioner uploads a local state tree, which is required. This feature gives the option to download community SaltStack formulas as an array of URLs (consumed by go-getter) inside the states directory *before* the state tree is uploaded to the destination. This allows operators using the provisioner to consume any formula on https://github.com/saltstack-formulas using git-style URLs. 

Example usage:
```json
{
  "provisioners": [
    {
      "type": "salt-masterless",
      "local_state_tree": "salt",
      "formulas": [
        "git::https://github.com/saltstack-formulas/vault-formula.git//vault",
        "git::https://github.com/saltstack-formulas/consul-formula.git//consul",
        "git::https://github.com/saltstack-formulas/consul-formula.git//consul-template"
      ]
    }
  ]
}
```
The result is something like:

```log
file: output will be in this color.

==> file: Downloading Salt formulas...
    file: git::https://github.com/tucows/vault-formula.git//vault => salt/vault
    file: git::https://github.com/tucows/consul-formula.git//consul => salt/consul
    file: git::https://github.com/tucows/consul-formula.git//consul-template => salt/consul-template
==> file: Provisioning with Salt...
    file: Downloading saltstack bootstrap to /tmp/install_salt.sh
    file: Installing Salt with command sudo sh /tmp/install_salt.sh
    file: Creating remote temporary directory: /tmp/salt
    file: Creating directory: /tmp/salt
    file: Uploading local state tree: salt
    file: Moving states to /tmp/salt/states
    file: Verifying Path: /srv/salt
    file: Moving /tmp/salt/states to /srv/salt
    file: Removing Salt formula: salt/vault
    file: Removing Salt formula: salt/consul
    file: Removing Salt formula: salt/consul-template
    file: Running: salt-call --local  state.highstate --file-root=/srv/salt --pillar-root=/srv/pillar --retcode-passthrough -l info
Build 'file' finished.

==> Builds finished. The artifacts of successful builds are:
--> file: Stored file: dummy_artifact
```

Prior to this, we were pulling in community formulas with the shell provisioner, which resulted in an identical script for each packer repo. The goal is to provide a more native experience in Packer to download salt formulas from external sources while still using salt-masterless.

Looking for feedback, as this is my first PR into Packer 😃 . As far as tests are concerned, the salt-masterless provisioner only really has `Prepare` tests, so I stayed consistent with that.

Thanks for your time.

### CODE REVIEW

**Comment by SwampDragons** [comment]
Hi! Welcome to the community and thanks for such a cool addition. This is a nice PR. I figured I'd save you a couple cycles by fixing the linting and imports for you; gonna push a few commits to your fork to get tests green.

**Comment by SwampDragons** [comment]
One thing I'd like you to do though -- add some docs for this new feature to https://github.com/hashicorp/packer/blob/master/website/pages/docs/provisioners/salt-masterless.mdx for me?

**Comment by acornies** [comment]
Certainly 👍 

I've also identified an update to the regexp so operators can specify a `?ref=` in git urls for pinning.

File: go.mod [hunk 1]
```diff
@@ -59,6 +59,7 @@ require (
 	github.com/hashicorp/go-checkpoint v0.0.0-20171009173528-1545e56e46de
 	github.com/hashicorp/go-cleanhttp v0.5.1
 	github.com/hashicorp/go-cty-funcs v0.0.0-20200520133146-0d04eb807361
+	github.com/hashicorp/go-getter v1.4.1
```
Comment (sylviamoss) [inline go.mod:62]: Packer uses the v2 version of `go-getter` which is already a dependency here.

File: provisioner/salt-masterless/provisioner.go [hunk 2]
```diff
@@ -11,8 +11,10 @@ import (
 	"fmt"
 	"os"
 	"path/filepath"
+	"regexp"
 	"strings"
 
+	"github.com/hashicorp/go-getter"
```
Comment (sylviamoss) [inline provisioner/salt-masterless/provisioner.go:17]:
```suggestion
	"github.com/hashicorp/go-getter/v2"
```

Comment (acornies) [reply provisioner/salt-masterless/provisioner.go]: Good point! I've updated the implementation to use go-getter/v2, did the go mod/vendor dance.

Comment (sylviamoss) [reply provisioner/salt-masterless/provisioner.go]: Nice!! Thanks!

**Comment by SwampDragons** [comment]
Thanks!

**Pull request merged**